### PR TITLE
[Quant][Inductor] Use truncate instead of default rounding round when convert float to uint8

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105456
* #105455
* #104590
* #104588
* #104581
* #104580
* #104573
* #105039
* __->__ #105109

**Summary**
When convert float tensor to uint8 data type as `tensor.to(dtype=torch.uint8)`, PyTorch will directly truncate the decimal. Previously, in `convert_float_to_uint8` we use `_mm512_cvtps_epi32` which uses default rounding mode (round to nearest) to convert float to uint8 which doesn't align with the eager mode behavior. Change  `_mm512_cvtps_epi32` to `_mm512_cvttps_epi32` to use directly truncate when convert float tensor to uint8.

**Test Plan**
```
python -m pytest test_cpu_repro.py -k test_to_uint8_rounding_method
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Comment by leslie-fang-intel** [comment]
Hi @jgong5 @mingfeima, could you help to take a look of this PR?

**Review by jgong5 — CHANGES_REQUESTED** [review]
I'm concerned about the accuracy impact of turning RNE to truncating. According to my previous experiences, RNE brings better overall accuracy than truncating rounding mode. Is it possible to apply RNE to `tensor.to`?

**Comment by leslie-fang-intel** [comment]
> I'm concerned about the accuracy impact of turning RNE to truncating. According to my previous experiences, RNE brings better overall accuracy than truncating rounding mode. Is it possible to apply RNE to `tensor.to`?

I have a discussion with @mingfeima yesterday. We think it should be expected that `tensor.to` will truncate instead of RNE when convert float to int. Also for the `to_dtype` node in inductor scalar path code gen, it will also generate code as `auto tmp1 = static_cast<unsigned char>(tmp0);`. Per my experiment, it will also truncate the float to int.

As for the accuracy, we think user needs to call `tensor.round` first before `tensor.to(uint8)` like the implementation of `decomposed.quantize_per_tensor`.
https://github.com/pytorch/pytorch/blob/4328138c1e8c90645fd68235299e0eb7f6d28cfd/torch/ao/quantization/fx/_decomposed.py#L62-L63

**Comment by jgong5** [comment]
> As for the accuracy, we think user needs to call `tensor.round` first before `tensor.to(uint8)` like the implementation of `decomposed.quantize_per_tensor`.

That makes sense to assume `tensor.round` would be called before the truncating. Thanks for the explanation.

**Review by jgong5 — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
Hi @mingfeima, could you kindly help to take a look of this PR?

File: aten/src/ATen/cpu/vec/vec256/vec256_qint.h [hunk 1]
```diff
@@ -109,8 +109,8 @@ inline Vectorized<float> convert_uint8_to_float(at::vec::Vectorized<uint8_t> src
 }
 
 inline Vectorized<uint8_t> convert_float_to_uint8(at::vec::Vectorized<float> src) {
-  // Convert from float32 to int32
-  __m256i x_values_int32 = _mm256_cvtps_epi32(src);
+  // Convert from float32 to int32, directly trunc
```
Comment (mingfeima) [inline aten/src/ATen/cpu/vec/vec256/vec256_qint.h:112]: the wording should be `Convert from float32 to int32 with truncation`

**Review by mingfeima — APPROVED** [review]
LGTM

File: aten/src/ATen/cpu/vec/vec256/vec256_qint.h (hunk 1 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec256/vec256_qint.h:112]: Thanks for the suggestion and changed.

**Comment by leslie-fang-intel** [comment]
Hi @eellison @jansel, could you kindly help to take a look of this PR? Looks need your approve to merge it.

**Review by jerryzh168 — APPROVED** [review]

**Pull request closed without merging**